configs/components/ruby-2.1.9.rb

```diff
@@ -5,21 +5,9 @@
   pkg.mirror "#{settings[:buildsources_url]}/ruby-#{pkg.get_version}.tar.gz"
```

Comment:
There's some shared code here and in ruby-2.4.2, from what I can find here. If you're interested, I have some ideas on how to clean this stuff up. 

https://github.com/caseywilliams/puppet-runtime/blob/agent-runtime/configs/components/ruby-2.1.9.rb#L7-L9
https://github.com/caseywilliams/puppet-runtime/blob/agent-runtime/configs/components/ruby-2.4.2.rb#L7-L9

Common patches for AIX and Windows

This part:
https://github.com/caseywilliams/puppet-runtime/blob/agent-runtime/configs/components/ruby-2.1.9.rb#L127-L130
https://github.com/caseywilliams/puppet-runtime/blob/agent-runtime/configs/components/ruby-2.4.2.rb#L128L131

This part:
https://github.com/caseywilliams/puppet-runtime/blob/agent-runtime/configs/components/ruby-2.1.9.rb#L132-L142
https://github.com/caseywilliams/puppet-runtime/blob/agent-runtime/configs/components/ruby-2.4.2.rb#L133-L144

(one has an extra zlib-devel on AIX, aside from that things are the same)

This part:
https://github.com/caseywilliams/puppet-runtime/blob/agent-runtime/configs/components/ruby-2.1.9.rb#L144-L150
https://github.com/caseywilliams/puppet-runtime/blob/agent-runtime/configs/components/ruby-2.4.2.rb#L146-L152

Most of these parts:
https://github.com/caseywilliams/puppet-runtime/blob/agent-runtime/configs/components/ruby-2.1.9.rb#L152-L275
https://github.com/caseywilliams/puppet-runtime/blob/agent-runtime/configs/components/ruby-2.4.2.rb#L171-L291

Reply:
Please do let me know your suggestions

Reply:
Speaking of ruby 2.4.2, it's worth noting in this PR that there's are open PRs to the agent that update ruby 2.4.2 and patch ruby 2.1.9 - I will need to update this once those are merged.

Reply:
It's something like an instance_eval at some point in the middle or the end since here, you need to have setup things like the `rbconfig_info` + the `base` directory for patches so you can't do it at the beginning like you could w/ the `rubygem` stuff.

In the component files, you would declare the `rbconfig` variable and the custom `base` dir variable, the latter is custom since you will probably be applying some common patches (e.g. for AIX) in the base component file, whereas `rbconfig` is structurally the same. From there, you can put the common stuff in the base component file and then call an instance_eval, where the base component file's like what you have for `rubygem-base`.

Only fault with that is you lose the nice structure that the current component files give you (apply patches, setup environments + requirements, then set up installation command). If you'd like that structure, then you can still do instance_eval but just have methods for those specific steps (`setup_common_environment`) in the base file that you can call when you need in the specialized files (e.g. `setup_base_environment` => apply environment overrides).